### ISSUE TITLE
parallel: Control default timeout and setup_timeout for PSOCK clusters

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/Wishlist-for-R#91]
# Background

The default values for `timeout` and `setup_timeout` arguments for `parallel::makePSOCKcluster()` are defined in [src/library/parallel/R/snow.R](https://github.com/wch/r-source/blob/R-3-5-branch/src/library/parallel/R/snow.R#L102-L103):
```r
initDefaultClusterOptions <- function(libname)
{
  ...
                    setup_timeout = 60 * 2,      # 2 minutes
                    timeout = 60 * 60 * 24 * 30, # 30 days
  ...
```

where:

* `setup_timeout` - the maximum time master and background PSOCK workers will wait for each other to establish a socker connection.
* `timeout` - the maximum time the PSOCK workers will wait for the next "command" from the master.

# Issue

If a package test sets up a PSOCK cluster and then the master process dies for one reason or the other, the PSOCK worker processes will remain running for 30 days ('timeout') until they timeout and terminate that way.  When this happens on CRAN servers, where many packages are checked all the time, this will result in a lot of stray R processes.

## Example

Here is an example illustrating how R leaves behind stray R processes if fails to establish a connection to one or more background R processes launched by `parallel::makeCluster()`;

First, let's show that there are no R processes running:
```sh
$ ps aux | grep -E "exec[/]R"
```

Then, lets create a PSOCK cluster for which connection will fail (because port 80 is reserved):
```sh
$ Rscript -e 'parallel::makeCluster(1L, port=80)'
Error in socketConnection("localhost", port = port, server = TRUE, blocking = TRUE,  : 
  cannot open the connection
Calls: <Anonymous> ... makePSOCKcluster -> newPSOCKnode -> socketConnection
In addition: Warning message:
In socketConnection("localhost", port = port, server = TRUE, blocking = TRUE,  :
  port 80 cannot be opened
```

The launched R worker is still running:
```sh
$ ps aux | grep -E "exec[/]R"
hb       20778 37.0  0.4 283092 70624 pts/0    S    17:50   0:00 /usr/lib/R/bin/exec/R --slave --no-restore -e parallel:::.slaveRSOCK() --args MASTER=localhost PORT=80 OUT=/dev/null SETUPTIMEOUT=120 TIMEOUT=2592000 XDR=TRUE
```

This process will keep running for `TIMEOUT=2592000` seconds (= 30 days).  The reason for this is that it is currently in the state where it attempts to set up a connection to the main R process:

```r
> parallel:::.slaveRSOCK
function () 
{
    makeSOCKmaster <- function(master, port, setup_timeout, timeout, 
        useXDR) {
        port <- as.integer(port)
        timeout <- as.integer(timeout)
        stopifnot(setup_timeout >= 0)
        retryDelay <- 0.1
        retryScale <- 1.5
        t0 <- Sys.time()
        repeat {
            con <- tryCatch({
                socketConnection(master, port = port, blocking = TRUE, 
                  open = "a+b", timeout = timeout)
            }, error = identity)
    ...	    
}	    
```

In other words it is stuck in `socketConnection()` and it won't time out until `timeout' seconds.

# Suggestion

To mitigate the problem with above stray processes from running `R CMD check`, we could have the latter shorten the default `timeout`.  This could be done by making it possible to control the defaults via environment variables, e.g.

```r
setup_timeout = as.numeric(Sys.getenv("R_PARALLEL_SETUP_TIMEOUT", 60 * 2)),      # 2 minutes
timeout = as.numeric(Sys.getenv("R_PARALLEL_SETUP_TIMEOUT", 60 * 60 * 24 * 30)), # 30 days
```

This would allow CRAN and others to use, say,
```r
R_PARALLEL_SETUP_TIMEOUT=60
R_PARALLEL_SETUP_TIMEOUT=120
```

**Comment by HenrikBengtsson** [comment]
Posted a version of the above to R-devel thread ['SUGGESTION: Proposal to mitigate problem with stray processes left behind by parallel::makeCluster()'](https://stat.ethz.ch/pipermail/r-devel/2019-March/077477.html) on 2019-03-17.

**Comment by HenrikBengtsson** [comment]
This has been fixed in [R-devel r76275 (2019-03-27)](https://github.com/wch/r-source/commit/1cef346ef58b01e0e12b009824a2a317f1335c44).  This fix was as simple as:

```diff
-  if (Sys.time() - t0 > setup_timeout) break
+  if (difftime(Sys.time(), t0, units="secs") > setup_timeout) break
```

so my troubleshooting above was incorrect.
